storage/remote/otlptranslator/prometheusremotewrite/helper.go

```diff
@@ -56,182 +57,117 @@ const (
-	scopeLabelCount := 0
-	if promoteScope {
-		// Include name, version and schema URL.
-		scopeLabelCount = scope.attributes.Len() + 3
-	}
-
-	// Calculate the maximum possible number of labels we could return so we can preallocate l.
-	maxLabelCount := attributes.Len() + len(settings.ExternalLabels) + len(promotedAttrs) + scopeLabelCount + len(extras)/2
-
-	if haveServiceName {
-		maxLabelCount++
-	}
-	if haveInstanceID {
-		maxLabelCount++
-	}
-	if settings.EnableTypeAndUnitLabels {
-		maxLabelCount += 2
-	}
 
 	// Ensure attributes are sorted by key for consistent merging of keys which
 	// collide when sanitized.
-	labels := make([]prompb.Label, 0, maxLabelCount)
+	c.scratchBuilder.Reset()
+
 	// XXX: Should we always drop service namespace/service name/service instance ID from the labels
 	// (as they get mapped to other Prometheus labels)?
 	attributes.Range(func(key string, value pcommon.Value) bool {
 		if !slices.Contains(ignoreAttrs, key) {
-			labels = append(labels, prompb.Label{Name: key, Value: value.AsString()})
+			c.scratchBuilder.Add(key, value.AsString())
 		}
 		return true
 	})
-	sort.Stable(ByLabelName(labels))
+	c.scratchBuilder.Sort()
+	sortedLabels := c.scratchBuilder.Labels()
 
-	// map ensures no duplicate label names.
-	l := make(map[string]string, maxLabelCount)
 	labelNamer := otlptranslator.LabelNamer{UTF8Allowed: settings.AllowUTF8}
-	for _, label := range labels {
-		finalKey, err := labelNamer.Build(label.Name)
-		if err != nil {
-			return nil, err
-		}
-		if existingValue, alreadyExists := l[finalKey]; alreadyExists {
-			l[finalKey] = existingValue + ";" + label.Value
-		} else {
-			l[finalKey] = label.Value
+
+	if settings.AllowUTF8 {
+		// UTF8 is allowed, so conflicts aren't possible.
+		c.builder.Reset(sortedLabels)
+	} else {
+		// Now that we have sorted and filtered the labels, build the actual list
+		// of labels, and handle conflicts by appending values.
+		c.builder.Reset(labels.EmptyLabels())
+		var sortErr error
+		sortedLabels.Range(func(l labels.Label) {
+			finalKey, err := labelNamer.Build(l.Name)
+			if err != nil && sortErr == nil {
+				sortErr = err
+				return
+			}
+			if existingValue := c.builder.Get(finalKey); existingValue != "" {
+				c.builder.Set(finalKey, existingValue+";"+l.Value)
+			} else {
+				c.builder.Set(finalKey, l.Value)
+			}
+		})
+		if sortErr != nil {
+			return labels.EmptyLabels(), sortErr
 		}
 	}
 
-	for _, lbl := range promotedAttrs {
-		normalized, err := labelNamer.Build(lbl.Name)
-		if err != nil {
-			return nil, err
+	if settings.EnableTypeAndUnitLabels {
+		unitNamer := otlptranslator.UnitNamer{UTF8Allowed: settings.AllowUTF8}
+		if meta.Type != model.MetricTypeUnknown {
+			c.builder.Set("__type__", strings.ToLower(string(meta.Type)))
 		}
-		if _, exists := l[normalized]; !exists {
-			l[normalized] = lbl.Value
+		if meta.Unit != "" {
+			c.builder.Set("__unit__", unitNamer.Build(meta.Unit))
 		}
 	}
```

Comment:
This changes the order of operations and hence behaviour AFAICT, and I don't see any good reason for it:
```suggestion
	for _, lbl := range promotedAttrs {
		normalized, err := labelNamer.Build(lbl.Name)
		if err != nil {
			return nil, err
```